src/goto-programs/remove_virtual_functions.cpp

```diff
@@ -209,15 +216,53 @@ void remove_virtual_functionst::remove_virtual_function(
       t3->make_goto(t_final, true_exprt());
     }
 
+    // Emit target if end of dispatch table is reached or if the next element is
+    // dispatched to another function call. Assumes entries in the functions
+    // variable to be sorted for the identifier of the function to be called.
+    auto l_it = std::next(it);
+    bool next_emit_target = (l_it == functions.crend());
+    if(!next_emit_target)
+    {
+      next_emit_target =
+        l_it->symbol_expr.get_identifier() != fun.symbol_expr.get_identifier();
+    }
+
+    // The root function call is done via fall-through, so nothing to emit
+    // explicitly for this.
+    if(next_emit_target && fun.symbol_expr == last_function_symbol)
+    {
+      class_ids.clear();
+    }
+
     // If this calls the fallback function we just fall through.
     // Otherwise branch to the right call:
     if(fallback_action!=virtual_dispatch_fallback_actiont::CALL_LAST_FUNCTION ||
        fun.symbol_expr!=last_function_symbol)
     {
-      exprt c_id1=constant_exprt(fun.class_id, string_typet());
-      goto_programt::targett t4=new_code_gotos.add_instruction();
-      t4->source_location=vcall_source_loc;
-      t4->make_goto(insertit.first->second, equal_exprt(c_id1, c_id2));
+      // create a disjunction of class_ids to test
+      if(next_emit_target && fun.symbol_expr != last_function_symbol)
+      {
+        exprt::operandst or_ops;
+        for(const auto &id : class_ids)
+        {
+          exprt c_id1 = constant_exprt(id, string_typet());
+          equal_exprt class_id_test(c_id1, c_id2);
+          or_ops.push_back(class_id_test);
+        }
+
+        goto_programt::targett t4 = new_code_gotos.add_instruction();
+        t4->source_location = vcall_source_loc;
+        t4->make_goto(insertit.first->second, disjunction(or_ops));
```

Comment:
I find the variable name `t4` not very descriptive, I would suggest something like `target_for_goto_instruction`.